### ISSUE TITLE
Container identification should be service rather than image

### ISSUE THREAD

**Issue by navneetkarnani** [microsoft/vscode-remote-release#4237]
Issue Type: <b>Bug</b>

I am using Remote Containers to connect to my development environment that is hosted using docker-compose, to avoid having to download the toolchain versions to the local machine. 

The Remote Containers extension, when connecting to the container, uses the 'image' as the unique identification for setting. This results in all the services that I am developing getting the same dev settings, including the tree mount point and the extensions. 

eg. My docker-compose has three services, db (postgres), and two java development for api, and service ( maven:3.6-openjdk-15 ) containers. have volumes mounted for code, and home directories. When I connect to api, and install extensions, and open the tree to point to /usr/src/api, all works fine. Then, when I connect to  service, I see the tree mounted on /usr/src/api and the extensions already installed. 

The expected behaviour is that the two environments should get clean setups and work from there, without interfering with each other.

Extension version: 0.154.1
VS Code version: Code 1.52.1 (ea3859d4ba2f3e577a159bc91e3074c5d85c0523, 2020-12-16T16:30:02.420Z)
OS version: Darwin x64 20.2.0
Remote OS version: Linux x64 4.19.121-linuxkit
Remote OS version: Linux x64 4.19.121-linuxkit

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz (4 x 1600)|
|GPU Status|2d_canvas: enabled<br>flash_3d: enabled<br>flash_stage3d: enabled<br>flash_stage3d_baseline: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|3, 3, 3|
|Memory (System)|16.00GB (0.23GB free)|
|Process Argv|. --crash-reporter-id 0dec9fd0-c8fd-4152-a33c-72a9d4647e00|
|Screen Reader|no|
|VM|0%|

|Item|Value|
|---|---|
|Remote|Container maven:3.6-openjdk-15 (/mocsh-dev_webapi_1)|
|OS|Linux x64 4.19.121-linuxkit|
|CPUs|Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz (2 x 1600)|
|Memory (System)|3.85GB (0.11GB free)|
|VM|0%|

|Item|Value|
|---|---|
|Remote|Container maven:3.6-openjdk-15 (/mocsh-dev_gmail-service_1)|
|OS|Linux x64 4.19.121-linuxkit|
|CPUs|Intel(R) Core(TM) i5-8210Y CPU @ 1.60GHz (2 x 1600)|
|Memory (System)|3.85GB (0.11GB free)|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv368:30146709
vsreu685:30147344
openlogontheside:30221877
python763:30178808
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
python504:30227505
vswsl492cf:30211402
wsl2prompt:30224612
pythonvsdeb440:30237827
unusedpromptcf:30224611
folderexplorercf:30224615
openfilemenucf:30224648
pythonvsded773cf:30236630
vsjup459cf:30237188

```

</details>

**Comment by PavelSosin-320** [comment]
You are right, the container can't be identified by its image, the relation is  1..n*
The container's unique identification is Docker's Container ID. Container can't be identified by the Service name due to Service scalability.
The container can be labeled. It must be labeled to use a match pattern in container management. There is no inverse reference from the container to the service, only filter.
Multiple labels are supported via --label and --label-file options.

**Comment by PavelSosin-320** [comment]
@navneetkarnani When you run docker-compose --help the 1st line of the help is 
"Define and run multi-container applications with Docker." The topmost section of the compose-file is SERVICE.
You can use --compatibility option to use the default non-swarm scaling. Anyway, the recent docker-compose release looks at the docker-context and if it sees orchestration it will make its conclusion.

**Comment by navneetkarnani** [comment]
@PavelSosin-320 My problem was not multiple contexts. 
I found a workaround to the problem, listed deep inside the Remote Containers extension documentation, under the "named configurations" title.

The problem is that all the images in a "development" environment may come from the same image. So, even though I want them to have different settings, because the Remote Containers extension uses the image as the configuration base, all the containers get the same settings, and then there is a continuous conflict.

The Solution is outlined here: https://code.visualstudio.com/docs/remote/attach-container#_attached-container-configuration-files

My steps to get around my specific reported problem: 
Copy the json file from the `imageConfigs` directory into the `nameConfig` directory, under the `~/Library/Application Support/Code/User/globalStorage/ms-vscode-remote.remote-containers`, with the name of the container that I have created under docker.

The documentation for this is is not clearly visible, or even easily reachable. Everytime I want to find it, I have to go down multiple paths, before I can locate it. And that too, because I know what I am looking for.

It would be great to have place in the settings where the current "imageConfigs" would be listed, and a way to `bind` it to a named container.

**Comment by PavelSosin-320** [comment]
In any workaround the docker expects container id as a parameter of docker attach. Using a container name makes the situation only worst. The primary purpose of Docker is to create as many container instances from the single image as Docker workload needs. Implementation of 1..1 relation between image and container instance contradicts Docker concept. Multiple running containers can share the same configuration. 
According to documentation the purpose of --name is not the unique identification but only  "If you specify a name, you can use it when referencing the container within a Docker network". This is DNS name of the container!!! Docker doesn't require name provisioning at all.
Docker always uses "match label" pattern to connect artifacts. If the configuration is labeled as "configuration=Config1" and Container is labeled as "configuration=Config1" these 2 object instances are related, up to the developer to define the cardinality.  The container can be easily found using docker container ls with --filter.

**Comment by navneetkarnani** [comment]
I'm sorry, but I think we are talking of different things here. My report is specifically for the "Remote Containers" extension of VS Code, and its handling of the "Attach to container" functionality.

**Comment by PavelSosin-320** [comment]
@navneetkarnani  Attach to container is the invocation of Docker's API call Docker container attach. Docker engine is not dedicated to run dev-containers. The role of dev-container extension code is very limited.
Neither using container image name due to its 1..n* relation cardinality with containers nor container name provided in the dockerfile or run command argument [ The container name meaning](https://www.digitalocean.com/community/tutorials/3-tips-for-namitg-docker-containers#:~:text=When%20you%20create%20a%20Docker,to%20the%20host%20and%20network.) and uniqueness  limitations is not so good idea Container name purpose is Display or DNS resolution. It is very limited by its format. The Docker engine knows nothing about files on the Host machine. Even less about files on the client machine.To create relation between running container and configuration artifact (file) the Container's metadata stored by the Docker engine has to be in the format of full or partial URL. It is possible if it is label value. Using label value Container can "point" to the file containing its configuration.

**Comment by chrmarti** [comment]
Related: https://github.com/microsoft/vscode-remote-release/issues/4013
